src/core/CurveDiscretizer.cc

```diff
@@ -0,0 +1,400 @@
+    auto current = q.top();
+
+    // Group similar length segmented edges to keep result roughly symmetrical.
+    while (!q.empty() && (tmp_q.empty() || q.top().close_match(tmp_q.front()))) {
+      tmp_q.push_back(q.top());
+      q.pop();
+    }
+
+    if (seg_total + tmp_q.size() <= fn) {
+      while (!tmp_q.empty()) {
+        current = tmp_q.back();
+        tmp_q.pop_back();
+        ++current.segment_count;
+        ++segment_counts[current.edge_index];
+        ++seg_total;
+        q.push(current);
+      }
+    } else {
+      // fn too low to segment last group, push back onto queue without change.
+      while (!tmp_q.empty()) {
+        current = tmp_q.back();
+        tmp_q.pop_back();
+        q.push(current);
+      }
+      break;
+    }
+  }
+
+  // Create final segmented edges.
+  Outline2d o2;
+  o2.positive = o.positive;
+  v0 = o.vertices[0];
+  for (size_t i = 1; i <= num_vertices; ++i) {
+    Vector2d v1 = o.vertices[i % num_vertices];
+    add_segmented_edge(o2, v0, v1, segment_counts[i - 1]);
+    v0 = v1;
+  }
+
+  assert(o2.vertices.size() <= fn);
+  return o2;
+}
+
+// For each edge in original outline, find its max length over all slice transforms,
+// and divide into segments no longer than fs.
+Outline2d splitOutlineByFs(const Outline2d& o, const double twist, const double scale_x,
+                           const double scale_y, const double fs, unsigned int slices)
+{
+  const auto num_vertices = o.vertices.size();
+
+  Vector2d v0 = o.vertices[0];
+  Outline2d o2;
+  o2.positive = o.positive;
+
+  // non-uniform scaling requires iterating over each slice transform
+  // to find maximum length of a given edge.
+  if (scale_x != scale_y) {
+    for (size_t i = 1; i <= num_vertices; ++i) {
+      Vector2d v1 = o.vertices[i % num_vertices];
+      double max_edgelen = 0.0;  // max length for single edge over all transformed slices
+      for (unsigned int j = 0; j <= slices; j++) {
+        double t = static_cast<double>(j) / slices;
+        Vector2d scale(Calc::lerp(1, scale_x, t), Calc::lerp(1, scale_y, t));
+        double rot = twist * t;
+        Eigen::Affine2d trans(Eigen::Scaling(scale) * Eigen::Affine2d(rotate_degrees(-rot)));
+        double edgelen = (trans * v1 - trans * v0).norm();
+        max_edgelen = std::max(max_edgelen, edgelen);
+      }
+      auto edge_segments = static_cast<unsigned int>(std::ceil(max_edgelen / fs));
+      add_segmented_edge(o2, v0, v1, edge_segments);
+      v0 = v1;
+    }
+  } else {  // uniform scaling
+    double max_scale = std::max(scale_x, 1.0);
+    for (size_t i = 1; i <= num_vertices; ++i) {
+      Vector2d v1 = o.vertices[i % num_vertices];
+      unsigned int edge_segments =
+        static_cast<unsigned int>(std::ceil((v1 - v0).norm() * max_scale / fs));
+      add_segmented_edge(o2, v0, v1, edge_segments);
+      v0 = v1;
+    }
+  }
+  return o2;
+}
+
+Outline2d CurveDiscretizer::splitOutline(const Outline2d& o, double twist, double scale_x,
```

Comment:
I'm not convinced this function belongs here, but let's leave that for another refactoring job.